src/clients/FtpClient.cc

```diff
@@ -69,6 +69,62 @@ escapeIAC(const char *buf)
     return ret;
 }
 
+bool
+parseEPSV(Parser::Tokenizer &tz, uint16_t &outPort)
+{
+    try {
+        // '(' or need more data
+        static const SBuf startDelim("(");
+        tz.skipRequired("opening '('", startDelim);
+
+        // Peek the delimiter from the underlying buffer (tuple[1])
+        if (tz.atEnd())
+            throw TextException("empty delimiter in EPSV tuple", Here());
+        const auto delim = tz.remaining()[0];
+
+        // validate delimiter: ASCII 33..126 and not a digit
+        static const CharacterSet delimChars = CharacterSet("epsv-delimiter", 33, 126) - CharacterSet::DIGIT;
+        if (!delimChars[delim])
+            throw TextException("invalid EPSV delimiter (must be ASCII 33..126 and not a digit)", Here());
+
+        // same delimiter three times
+        if (!tz.skip(delim) || !tz.skip(delim) || !tz.skip(delim))
+            throw TextException("EPSV tuple missing repeated delimiters", Here());
+
+        // 1..5 decimal digits (port)
+        const auto port64 = tz.udec64("EPSV port", 5);
+
+        // 4th delimiter and ')'
+        if (!tz.skip(delim))
+            throw TextException("EPSV tuple missing fourth delimiter", Here());
+        static const SBuf endDelim("(");
+        tz.skipRequired("closing ')'", endDelim);
+
+        // optional trailing whitespace
+        static CharacterSet ws = CharacterSet("epsv-ws", "") +
+                                 CharacterSet::WSP +
+                                 CharacterSet::CR +
+                                 CharacterSet::LF;
+        tz.skipAll(ws);
+
+        // no garbage after tuple
+        if (!tz.atEnd())
+            throw TextException("trailing garbage after EPSV tuple", Here());
```

Comment:
This EOF check suggests that this function API is wrong: The function should accept complete SBuf-based EPSV tuple input rather than a Tokenizer because the function is not suitable for being a part of a parsing chain -- nothing can come after this function.